src/models/token.rs

```diff
@@ -95,6 +97,30 @@ impl ApiToken {
         .or_else(|_| tokens.select(ApiToken::as_select()).first(conn))
         .map_err(Into::into)
     }
+
+    /// Find all tokens that are not revoked and will expire within the specified number of days.
+    pub fn find_tokens_expiring_within_days(
+        conn: &mut PgConnection,
+        days_until_expiry: i64,
+    ) -> QueryResult<Vec<ApiToken>> {
+        use diesel::dsl::{now, IntervalDsl};
+
+        api_tokens::table
+            .filter(api_tokens::revoked.eq(false))
+            .filter(
+                api_tokens::expired_at
+                    .is_not_null()
+                    .and(api_tokens::expired_at.assume_not_null().gt(now)) // Ignore already expired tokens
+                    .and(
+                        api_tokens::expired_at
+                            .assume_not_null()
+                            .lt(now + days_until_expiry.days()),
+                    ),
+            )
+            .filter(api_tokens::expiry_notification_at.is_null())
```

Comment:
hmm, given the name of the function I wouldn't expect this additional filter to be here. I understand why it's there for this specific purpose, but it makes the function less reusable. I'm wondering if we should move the function closer to the code that calls it, instead of putting it on the `ApiToken` impl block. 🤔

Reply:
I have moved it out in [53f70bf (#8290)](https://github.com/rust-lang/crates.io/pull/8290/commits/53f70bf814fe606526d4ca14c743f9efc04801ac).